binary_search_tree/binary_search_tree.go

```diff
@@ -0,0 +1,198 @@
+			n.Replace(node, n.Left)
+		} else {
+			replacement, parent := n.Left.Max(n)
+			n.Key = replacement.Key
+			n.Value = replacement.Value
+			replacement.Delete(replacement.Key, parent)
+		}
+	}
+}
+
+// BST (Binary Search Tree) struct just contains the root node
+// so client code doesnt have to work with the nodes
+type BST struct {
+	root *node
+}
+
+// Insert adds an element to the BST
+// Insertion is O(log n)
+func (bst *BST) Insert(key, value int) {
+	if bst.root == nil {
+		// if root is nil, i.e. the tree is empty,
+		// just assign new node to root
+		bst.root = &node{
+			Key:   key,
+			Value: value,
+		}
+	} else {
+		// otherwise call insert on root node
+		bst.root.Insert(key, value)
+	}
+}
+
+// Search traverses the tree trying to find a node
+// with given key. If successfull, the value of the node
+// will be returned with a true value. Otherwise 0 and false
+// Search is O(log n)
+func (bst *BST) Search(key int) (int, bool) {
+	// start at root and traverse tree
+	for node := bst.root; node != nil; {
+		if key == node.Key {
+			// if keys are equal: success
+			return node.Value, true
+		}
+		if key < node.Key {
+			// if key is lower, continue loop with left node
+			node = node.Left
+		} else {
+			// if key is higher, continue loop with right node
+			node = node.Right
+		}
+	}
+	return 0, false
+}
+
+// Delete removes a node in the tree with the corresponding key
+func (bst *BST) Delete(key int) {
+	if bst.root != nil {
+		// create fake parent node
+		parent := &node{Right: bst.root}
+		// call delete on root node
+		bst.root.Delete(key, parent)
+	}
+}
+
+// Traverse traverses the tree in order (https://en.wikipedia.org/wiki/Tree_traversal#In-order)
+// and calls the supplied function for each node
+func (bst *BST) Traverse(f func(key, value int)) {
+	// call helper function on root
+	traverse(bst.root, f)
+}
+
+// helper for traverse
+func traverse(node *node, f func(key, value int)) {
+	if node == nil {
+		// stop if given node is nil
+		return
+	}
+	// traverse down left subtree
+	traverse(node.Left, f)
+	// call function on node
+	f(node.Key, node.Value)
+	// traversen down right subtree
+	traverse(node.Right, f)
+}
+
+func main() {
+	// create empty bst
+	var bst BST
+	// add some key value pairs
+	bst.Insert(8, 9)
+	bst.Insert(4, 2)
+	bst.Insert(12, 19)
+	bst.Insert(2, 8)
+	bst.Insert(1, 200)
+	bst.Insert(10, 16)
+
+	// Print in order. Should print from lowest key to highest
+	bst.Traverse(func(key, value int) {
+		fmt.Printf("Key: %d, Value: %d\n", key, value)
+	})
+
+	// Searching for a key. Should return value and whether key was found or not
+	value, ok := bst.Search(12)
+	if ok {
+		fmt.Printf("Key 12 was found! It's value is %d\n", value)
+	}
+
+	// Delet a key
+	bst.Delete(12)
+	// Search for same key again
+	_, ok = bst.Search(12)
+	if ok {
+		fmt.Println("This shouldn't happen, since key should have been deleted")
+	} else {
+		fmt.Println("12 was not found :)")
```

Comment:
Why it's "12 was not found". Use a variable if you want to use it more than once in a single code. Avoid hardcoding.